src/main/scala/nl/biopet/tools/multigenicsearch/MultigenicSearch.scala

```diff
@@ -0,0 +1,196 @@
+import org.apache.spark.sql.expressions.UserDefinedFunction
+import org.apache.spark.sql.{Dataset, SparkSession}
+import org.apache.spark.sql.functions._
+
+import scala.collection.JavaConversions._
+import scala.concurrent.{Await, Future}
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration.Duration
+
+object MultigenicSearch extends ToolCommand[Args] {
+  def emptyArgs = Args()
+  def argsParser = new ArgsParser(this)
+
+  def main(args: Array[String]): Unit = {
+    val cmdArgs = cmdArrayToArgs(args)
+
+    logger.info("Start")
+
+    implicit val spark: SparkSession = SparkSession.builder().getOrCreate()
+    implicit val sc: SparkContext = spark.sparkContext
+    import spark.implicits._
+
+    val reader = new VCFFileReader(cmdArgs.inputFile, true)
+    val header = reader.getFileHeader
+    reader.close()
+    val sampleNumber = header.getNGenotypeSamples
+
+    val scatter = BedRecordList.fromReference(cmdArgs.reference).scatter(1000000)
+
+    val scatterRegions = sc.parallelize(
+      scatter, scatter.size)
+    val variants = readVcf(scatterRegions, cmdArgs.inputFile, sampleNumber)
+
+    val indexOnly = variants
+      .select("index", "variant.samples")
+      .withColumnRenamed("samples", "variantSamples")
+      .as[SingleSamples]
+
+    val maxMismatches: Map[Int, Int] = (2 to cmdArgs.maxCombinationSize)
+      .map(i => i -> (i - (i * cmdArgs.multigenicFraction).ceil.toInt)).toMap
+    val maxMismatch = maxMismatches.values.max
+
+    val combinations = (3 to cmdArgs.maxCombinationSize)
+      .foldLeft(Map(2 -> filterMaxMismatches(initCombinations(indexOnly), maxMismatch, sampleNumber, cmdArgs.sampleFraction))) { (a,b) =>
+      val c = addCombination(indexOnly, a(b-1))
+      a + (b -> filterMaxMismatches(c, maxMismatch, sampleNumber, cmdArgs.sampleFraction))
+    }
+
+    val futures = (2 to cmdArgs.maxCombinationSize).map { i =>
+      writeResult(variants, combinations(i), maxMismatches(i), sampleNumber, cmdArgs.sampleFraction, new File(cmdArgs.outputDir, s"multigenic_$i"))
+    }
+
+    Await.result(Future.sequence(futures), Duration.Inf)
+
+    Thread.sleep(1000000)
+
+    spark.stop()
+    logger.info("Done")
+  }
+
+  def filterMaxMismatches(combinations: Dataset[Combination],
+                          maxMismatch: Int,
+                          sampleNumber: Int,
+                          sampleFraction: Double): Dataset[Combination] = {
+    combinations.filter { x =>
+      val mismatches = (0 until sampleNumber).map(s => x.samples.map(_(s)).count(_ == false))
+      val mismatchCount = mismatches.count(_ >= maxMismatch)
+      mismatchCount.toDouble / sampleNumber.toDouble >= sampleFraction
+    }
+  }
+
+  def readVcf(scatterRegions: RDD[List[BedRecord]],
+              inputFile: File,
+              sampleNumber: Int)(
+      implicit spark: SparkSession): Dataset[SingleVariantWithIndex] = {
+    import spark.implicits._
+    scatterRegions
```

Comment:
Maybe add some intermediate variables or comments in this workflow to make it more readable.